nabs/plan_stubs.py

```diff
@@ -60,3 +64,96 @@ def measure_average(detectors, num, delay=None, stream=None):
     yield from stub_wrapper(count(detectors, num=num, delay=delay))
     # Return the measured average as a dictionary for use in adaptive plans
     return stream.last_event
+
+
+def move_per_step(detectors, step, pos_cache):
+    """
+    The motion portion of the per_step hook.
+
+    Use this inside of other per_step hooks to avoid rewriting the move part.
+    """
+    yield from checkpoint()
+    grp = short_uid('set')
+    for motor, pos in step.items():
+        if pos == pos_cache[motor]:
+            continue
+        yield from abs_set(motor, pos, group=grp)
+        pos_cache[motor] = pos
+    yield from wait_uid(group=grp)
+
+
+def monitor_events(detectors, events=None, duration=None):
+    """
+    Monitor for value updates from detectors and wait for completion.
+
+    This can run for a fixed number of events, for a fixed duration, or both.
+    If both, this will run until both conditions are met: until enough time has
+    passed and until enough events have been seen.
+
+    If events is used, we'll wait until every detector reaches the set events
+    count.
+
+    Parameters
+    ----------
+    events: ``int``, optional
+        If provided, we'll monitor until we have enough events.
+    duration: ``float``, optional
+        If provided, we'll monitor for this fixed duration in seconds.
+        If both arguments are provided, then we'll wait for both to be true,
+        so we'll have a minimum elapsed time and a minimum number of events.
```

Comment:
Before I read the docstring I assumed you would `or` the two conditions. Not sure this is wrong just surprised. The bonus of having it `or` is you essentially use the `duration` as a timeout.

Reply:
Interesting, using `or` does make this more intuitive. It's `and` simply because that was the default behavior that fell out when I implemented it the first time, and I don't know how to set a timeout on the future wait. Maybe this would be better implemented by adding a `wait_status` message (something I've wanted for a while) so we have a reasonable way to `or` here.